### PR TITLE
Add cupy put

### PR DESCRIPTION
**Pull request by yuyu2172**
Merge after #1855 

This PR adds `cupy.put` that functions similarly to `numpy.put`, but different in the following way.
+ `numpy.put` handles values sequentially, whereas `cupy.put` does not. 

Here is an example that illustrates the difference.

```python
>>> a = xp.zeros((2,))

>>> i = xp.arange(10000) % 2
>>> v = xp.arange(10000).astype(np.float)

>>> xp.put(a, i, v)
```

On numpy, whenever there is duplicating indices, the index that appeared last is used to store a value.
```python
>>> print a  # when xp is numpy
[ 9998. 9999.]
```

Whereas, in my implementation of `cupy.put`, the duplicating indices do not get handled as Numpy does. Instead, the index that gets handled by the last called kernel thread that stores values.
```python
>>> print a  # when xp is cupy
[ 9982. 9983.]
```

This can be changed to be compatible to Numpy. However, it requires an extra kernel call to look for duplicates, which I am not sure if worth it.
I would like to hear feedback. Thanks.

##### UPDATE:
I decided to leave CuPy different from NumPy with a index of duplicating values for performance reason.

### CODE REVIEW

File: tests/cupy_tests/core_tests/test_ndarray.py [hunk 1]
```diff
@@ -181,3 +204,107 @@ def test_output_type_mismatch(self, xp):
...
+        v_shape = compute_v_shape(
+            self.shape, numpy.array(self.indices).shape, self.axis)
+        v = testing.shaped_arange(v_shape, xp, dtype)
+        wrap_put(a, self.indices, v, self.axis)
+        return a
+
+
+@testing.parameterize(
+    {'shape': (3, 4, 5), 'indices': (2,), 'axis': 0, 'v_shape': (5,)},
+)
```
Comment (delta2323) [inline tests/cupy_tests/core_tests/test_ndarray.py]: We need not to use parameterized test if there is one combination of parameters.

**Comment by delta2323** [comment]
Could you add test cases in which `axis` is `None` and some elements in indices are out of range?

**Comment by delta2323** [comment]
According to the document, if `v` is shorter than `ind`, it will be repeated as necessary. Are there test cases that checks this behavior?

File: cupy/core/core.pyx [hunk 2]
```diff
@@ -1921,6 +1956,43 @@ cpdef ndarray _take(ndarray a, indices, axis=None, ndarray out=None):
...
+    v_shape = lshape + ind.shape + rshape
+    v = broadcast_to(v, v_shape)  # this was in manipulation/dims.pyx
+
+    cdim = ind.size
+    rdim = internal.prod(rshape)
+    ind = ind.reshape(
+        (1,) * len(lshape) + ind.shape + (1,) * len(rshape))
+    if axis == 0 or axis is None:
+        _put_kernel_0axis(v, ind, rdim, index_range, a.reduced_view())
+        return
```
Comment (yuyu2172) [inline cupy/core/core.pyx]: Thanks you for comment.

File: cupy/core/core.pyx [hunk 3]
```diff
@@ -1921,6 +1956,43 @@ cpdef ndarray _take(ndarray a, indices, axis=None, ndarray out=None):
...
+    else:
+        _put_kernel(v, ind, cdim, rdim, adim, index_range, a.reduced_view())
+        return
```
Comment (yuyu2172) [inline cupy/core/core.pyx]: Thanks you for comment.

File: tests/cupy_tests/core_tests/test_ndarray.py (hunk 1 above)
Comment (yuyu2172) [inline tests/cupy_tests/core_tests/test_ndarray.py]: Thanks you for comment.

File: cupy/core/core.pyx [hunk 4]
```diff
@@ -594,7 +594,16 @@ cdef class ndarray:
         """
         return _take(self, indices, axis, out)
 
-    # TODO(okuta): Implement put
+    cpdef put(self, ind, v):
+        """Replaces specified elements of an array with given values.
+
+        .. seealso::
+           :func:`cupy.put` for full documentation,
+           :meth: `numpy.ndarray.put`
```
Comment (yuyu2172) [inline cupy/core/core.pyx]: Thanks you for comment.

File: tests/cupy_tests/core_tests/test_ndarray.py [hunk 5]
```diff
@@ -36,6 +36,29 @@ def wrap_take(array, *args, **kwargs):
     return array.take(*args, **kwargs)
 
 
+def wrap_put(a, ind, v, axis=None):
+    if get_array_module(a) == numpy:
```
Comment (yuyu2172) [inline tests/cupy_tests/core_tests/test_ndarray.py]: Thanks you for comment.

File: tests/cupy_tests/core_tests/test_ndarray.py [hunk 6]
```diff
@@ -36,6 +36,29 @@ def wrap_take(array, *args, **kwargs):
...
+        if axis is not None and a.ndim != 0:
+            axis = axis % a.ndim
```
Comment (yuyu2172) [inline tests/cupy_tests/core_tests/test_ndarray.py]: Thanks you for comment.

File: tests/cupy_tests/core_tests/test_ndarray.py [hunk 7]
```diff
@@ -36,6 +36,29 @@ def wrap_take(array, *args, **kwargs):
...
+            a.put(ind, v, mode='wrap')
+    else:
+        core._put(a, ind, v, axis)
+
+
+def compute_v_shape(in_shape, indices_shape, axis):
+    if axis is None or len(in_shape) == 0:
+        return indices_shape
+    else:
+        axis = axis % len(in_shape)
```
Comment (yuyu2172) [inline tests/cupy_tests/core_tests/test_ndarray.py]: Thanks you for comment.

File: cupy/core/core.pyx [hunk 8]
```diff
@@ -1921,6 +1956,43 @@ cpdef ndarray _take(ndarray a, indices, axis=None, ndarray out=None):
             a.reduced_view(), indices, cdim, rdim, adim, index_range, out)
 
 
+cpdef _put(ndarray a, ind, v, axis=None):
```
Comment (yuyu2172) [inline cupy/core/core.pyx]: I am planning to use `_put` with `axis` when implementing `__setitem__` with array indexing in the future.

**Comment by yuyu2172** [comment]
@delta2323 
Thanks for the comments.

>Could you add test cases in which axis is None and some elements in indices are out of range and confirm it certainly raises error.

Actually, it will not raise error because current implementation fixes `mode` to `wrap` as done in `cupy.take`. 

> According to the document, if v is shorter than ind, it will be repeated as necessary. Are there test cases that checks this behavior?

I totally made mistake here. The documentation was a mistake. Such behavior is not supported.

**Comment by yuyu2172** [comment]
I start to feel that the functionality I would like to add `x[:, idx] = val` should not be called `put`, but something else.
I will resend PR with different name. 
One drawback is that no such method exists in `numpy`. I need to come up with the name of the method.

Sorry for inconvenience.
